Failed compiling LD2410

### Version

2025.8.0

### What type of installation are you using?

Home Assistant Add-on

### Browser

Edge

### What happened?

```
INFO ESPHome 2025.8.0
INFO Reading configuration /config/esphome/presencechambre.yaml...
INFO Detected timezone 'Europe/Paris'
INFO Generating C++ source...
INFO Compiling app...
Processing presencechambre (board: lolin_s2_mini; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/54.03.21-2/platform-espressif32.zip)
--------------------------------------------------------------------------------
INFO Package configuration completed successfully
INFO Package configuration completed successfully
HARDWARE: ESP32S2 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.50402.0 (5.4.2) 
 - tool-cmake @ 3.30.2 
 - tool-esp-rom-elfs @ 2024.10.11 
 - tool-esptoolpy @ 5.0.2 
 - tool-mklittlefs @ 3.2.0 
 - tool-ninja @ 1.13.1 
 - tool-scons @ 4.40801.0 (4.8.1) 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-xtensa-esp-elf @ 14.2.0+20241119
Reading CMake configuration...
Dependency Graph
|-- noise-c @ 0.1.10
|-- ArduinoJson @ 7.4.2
Compiling .pioenvs/presencechambre/src/esphome/components/ld2410/number/gate_threshold_number.cpp.o
Compiling .pioenvs/presencechambre/src/esphome/components/ld2410/number/max_distance_timeout_number.cpp.o
Compiling .pioenvs/presencechambre/src/esphome/components/ld2410/select/baud_rate_select.cpp.o
Compiling .pioenvs/presencechambre/src/esphome/components/ld2410/select/distance_resolution_select.cpp.o
src/esphome/components/ld2410/number/gate_threshold_number.cpp: In member function 'virtual void esphome::ld2410::GateThresholdNumber::control(float)':
src/esphome/components/ld2410/number/gate_threshold_number.cpp:10:18: error: 'class esphome::ld2410::LD2410Component' has no member named 'set_gate_threshold'
   10 |   this->parent_->set_gate_threshold(this->gate_);
      |                  ^~~~~~~~~~~~~~~~~~
src/esphome/components/ld2410/number/max_distance_timeout_number.cpp: In member function 'virtual void esphome::ld2410::MaxDistanceTimeoutNumber::control(float)':
src/esphome/components/ld2410/number/max_distance_timeout_number.cpp:8:18: error: 'class esphome::ld2410::LD2410Component' has no member named 'set_max_distances_timeout'
    8 |   this->parent_->set_max_distances_timeout();
      |                  ^~~~~~~~~~~~~~~~~~~~~~~~~
*** [.pioenvs/presencechambre/src/esphome/components/ld2410/number/gate_threshold_number.cpp.o] Error 1
*** [.pioenvs/presencechambre/src/esphome/components/ld2410/number/max_distance_timeout_number.cpp.o] Error 1
========================== [FAILED] Took 3.96 seconds ==========================
```

### How to reproduce

Create a component with LD2410

### Expected behavior

Compile must complete

### Relevant log output

```shell

```

### Screenshots

_No response_

--- a/esphome/config.py
+++ b/esphome/config.py
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
 import abc
+from collections import defaultdict
 from contextlib import contextmanager
 import contextvars
 import functools
@@ -37,11 +38,13 @@ from esphome.yaml_util import ESPForceValue, ESPHomeDataBase, is_secret
 _LOGGER = logging.getLogger(__name__)
 
 
-def iter_components(config):
+def iter_components(config: ConfigType, platform_counts: defaultdict[str, int]):
     for domain, conf in config.items():
         component = get_component(domain)
         yield domain, component
         if component.is_platform_component:
+            if not platform_counts.get(domain):
+                continue
             for p_config in conf:
                 p_name = f"{domain}.{p_config[CONF_PLATFORM]}"
                 platform = get_platform(domain, p_config[CONF_PLATFORM])

--- a/esphome/config.py
+++ b/esphome/config.py
@@ -43,7 +43,7 @@ def iter_components(config: ConfigType, platform_counts: defaultdict[str, int]):
         component = get_component(domain)
         yield domain, component
         if component.is_platform_component:
-            if not platform_counts.get(domain):
+            if not platform_counts[domain]:
                 continue
             for p_config in conf:
                 p_name = f"{domain}.{p_config[CONF_PLATFORM]}"

--- a/esphome/writer.py
+++ b/esphome/writer.py
@@ -181,7 +181,7 @@ the custom_components folder or the external_components feature.
 
 def copy_src_tree():
     source_files: list[loader.FileResource] = []
-    for _, component in iter_components(CORE.config):
+    for _, component in iter_components(CORE.config, CORE.platform_counts):
         source_files += component.resources
     source_files_map = {
         Path(x.package.replace(".", "/") + "/" + x.resource): x for x in source_files

